docker/scripts/download_genvm.sh

```
@@ -189,26 +191,40 @@ elif [[ -n "$REPO_ROOT" ]]; then
-    else
-        echo "WARNING: expected exactly one py-genlayer pin in Studio, found $RUNNER_PIN_COUNT; skipping runner assertion."
     fi
+    if [[ ! -d "$INSTALL_DIR/runners" ]]; then
+        error_exit "No runners/ directory after extracting $BUNDLE_DESC at $VERSION; check the published release assets."
+    fi
+    while read -r RUNNER_PIN; do
+        [[ -n "$RUNNER_PIN" ]] || continue
+        if ! RUNNER_TAR=$(runner_tar_for_pin "$RUNNER_PIN"); then
+            error_exit "Pinned py-genlayer runner is missing: py-genlayer:$RUNNER_PIN (looked in $INSTALL_DIR/runners and $INSTALL_DIR/executor/*/legacy-runners)."
+        fi
+        echo "Runner pin OK: py-genlayer:$RUNNER_PIN -> ${RUNNER_TAR#$INSTALL_DIR/}"
```

Shellcheck (0.11.0)

[info] Expansions inside ${..} need to be quoted separately, otherwise they match as patterns.

(SC2295)